src/main/scala/fs2/async/mutable/Queue.scala

```diff
@@ -0,0 +1,282 @@
+              // note resulting Seq[A] => F[Unit] is called whenever at least one `A` is available,
+              // however when more `A` are available then up to `sz` is returned
+              def registerDequeue(s:State[F,A],ref:F.Ref[Seq[A]], sz:Int):State[F,A] =
+                s.copy(deq = s.deq :+ (sz -> (F.setPure(ref)(_))))
+
+
+              // Tries to satisfy any awaiting de-queuers (recursively). Honors preferred size of dequeuer for chunking.
+              // Results in left, if there are remaining items after satisfying all the dequeuers
+              // and in right when input has been exhausted and there are still dequeuers available
+              // if, input has been fully consumed and all dequeuers satisfied emits empty on left
+              def fillDequeuer(xa:Seq[A], deq:Vector[DeQueue[F,A]]): F[Either[Seq[A],Vector[DeQueue[F,A]]]] = {
+                deq.headOption match {
+                  case None => F.pure(Left(xa))
+                  case Some((sz,dqF)) =>
+                    if (xa.isEmpty) F.pure(Right(deq))
+                    else {
+                      val (dqA, remA) = xa.splitAt(sz)
+                      F.bind(dqF(dqA))(_ => fillDequeuer(remA, deq.tail)) //todo: runs on single thread likely fork (S()) needed (dqF)? May we fork later in deque stream ?
+                    }
+                }
+              }
+
+              // recursively tries to satisfy deque operation so that acc contains up to `require` items
+              // it consults the `enq` to obtain more items for waiting enqueuers, while still respecting bound
+              // Results in tuple containing accumulated `A` with size <= require,  and enqueuers that were not used yet if any
+              // may return both empty, in that case queue is empty.
+              def satisfyDequeue(require:Int, enq:Vector[EnQueue[F,A]],acc:Vector[A]): F[(Vector[A],Vector[EnQueue[F,A]])] = {
+                if (acc.size >= require) F.pure(acc -> enq)
+                else {
+                  enq.headOption match {
+                    case None => F.pure(acc -> Vector.empty)
+                    case Some((xa,enqF)) =>
+                      val (nacc, rem) = (acc ++ xa).splitAt(require)
+                      if (rem.isEmpty) F.bind(enqF(0))(_ => satisfyDequeue(require,enq.tail, nacc))
+                      else F.pure((nacc,(rem,enqF) +: enq.tail))
+                  }
+                }
+              }
+
+              // Signals size change of queue, if that has changed
+              def signalSize(s:State[F,A], ns:State[F,A]) : F[Unit] = {
+                if (s.queue.size != ns.queue.size) szSignal.set(ns.queue.size)
+                else F.pure(())
+              }
+
+              // enqueueAll implementation
+              // tries to satisfy or waiting dequeuers on this enqueue and/or attaches the remainder to the queue
+              // also updates size of the queue.
+              def _enqueueAll(xa: Seq[A]): F[Int] = {
```

Comment:
I'd just implement this by repeatedly doing enqueueOne. The user can always let A be some chunk type if they want to cut down on some of the overhead.